hbase-asyncfs/src/main/java/org/apache/hadoop/hbase/io/asyncfs/FanOutOneBlockAsyncDFSOutputSaslHelper.java

```diff
@@ -502,12 +507,17 @@ private void checkSaslComplete() throws IOException {
       Set<String> requestedQop =
         ImmutableSet.copyOf(Arrays.asList(saslProps.get(Sasl.QOP).split(",")));
       String negotiatedQop = getNegotiatedQop();
+      // Treat null negotiated QOP as "auth" for the purpose of verification
+      // Code elsewhere does the same implicitly
+      if (negotiatedQop == null) {
+        negotiatedQop = "auth";
+      }
```

Comment:
I don't see why is this necessary here. It's only effective within this private method and doesn't make any difference in the verification at line 517, while it hides that there was no negotiated QoP with the client.

Since rest of the code handles "auth" and null equally, it'd make sense to return "auth" by the `getNegotiatedQop()` if null was negotiated effectively making sure that negotiatedQop will never be null. That would probably make some of the code in this class simpler, but still not strictly required for this patch.

Reply:
You are right, this only affects the logged values and what we return in the exception.

Reply:
However, having this makes logging consistent with HDFS.
I will open a ticket to address this in HDFS, and update the patch i if that is merged.

Reply:
That's not true.

The Hadoop code always requests some kind of QOP, at least "auth".

SASL mechanisms that don't support QOP at all, like SCRAM, will return null negotiated qop.

Without this if, we could not use SCRAM at all, as ["auth"] does not contain null.

The rest of the code does not check the negotiated QOP against the requested one, so a null check is fine there.

This is the simplest way I can think of to handle non-QOP capable SASL mechanisms.